src/Smalot/PdfParser/PDFObject.php

```diff
@@ -224,6 +224,65 @@ private function formatContent(?string $content): string
             return '';
         }
 
+        // Find all inline image content and replace them so they aren't
+        // affected by the next steps
+        $pdfInlineImages = [];
+        $offsetBI = 0;
+        while (preg_match('/\sBI\s(\/.+?)\sID\s(.+?)\sEI(?=\s|$)/s', $content, $text, \PREG_OFFSET_CAPTURE, $offsetBI)) {
+            // Attempt to detemine if this instance of the 'BI' command
+            // actually occured within a (string) using the following
+            // steps:
+
+            // Remove any escaped parentheses from the alleged image
+            // characteristics data
+            $para = str_replace(['\\(', '\\)'], '', $text[1][0]);
+
+            // Remove all correctly ordered and balanced parentheses
+            // from (strings)
+            do {
+                $paraTest = $para;
+                $para = preg_replace('/\(([^)]*)\)/', '$1', $paraTest);
+            } while ($para != $paraTest);
```

Comment:
Can we be sure this never can become an endless loop?

Reply:
The `preg_replace()` call is removing parentheses in pairs. Whenever it succeeds, `$para` is updated so it does not equal `$paraTest`. When it fails (does not find anything to replace), then `$para` remains the same and the loop exits. I can't see any way this falls into an endless loop.